### ISSUE TITLE
How can I change the kernel_divergence_fn property of a Bayesian layer during training?

### ISSUE THREAD

**Issue by nbro** [tensorflow/probability#887]
I need to change the `kernel_divergence_fn` property of each Bayesian layer of my model during training, so that it computes a slightly different thing depending on a certain input `k`. 

I am using Keras to fit the model, so I tried to implement a callback that gets the instance of the model, `k`, and I try to reassign to `kernel_divergence_fn` a new function that computes the KL divergence based on the value of `k`. This callback is called at the end of every batch.

However, this approach doesn't seem to take any effect, but it should because, according to the source code of [`DenseFlipout`](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/layers/dense_variational.py#L560-L722), `self.kernel_divergence_fn` is called in [the `call` method](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/layers/dense_variational.py#L180) (i.e. forward pass), so, every time we perform a forward pass, if I reassign `self.kernel_divergence_fn` at every batch, then `self.kernel_divergence_fn` should be different.

I am sure that my implementation of the function that computes the KL divergence based on `k` (i.e. `get_divergence_fn` below) is correct, because I can actually see its correct behaviour inside the callback 

```
def my_callback(model, k):
    def on_batch_end(batch, logs):
        for layer in model.layers:
            divergence_fn = get_divergence_fn(k) 
            
            # This prints the behaviour I am expecting, 
            # so get_divergence_fn and divergence_fn are correct.
            tf.print(divergence_fn(layer.kernel_posterior, layer.kernel_prior, None))
            
            # IT DOES NOT SEEM TO TAKE EFFECT
            layer.kernel_divergence_fn = divergence_fn
    
    return tf.keras.callbacks.LambdaCallback(on_batch_end=on_batch_end)
```

but the assignment

```
layer.kernel_divergence_fn = my_kernel_divergence_fn
```

does not seem to take effect, i.e. at the next epoch, `layer.kernel_divergence_fn` is still the one at the beginning of the epoch.

**Comment by bgroenks96** [comment]
It's inadvisable to rely on python level state changes when dealing with compiled Keras models. This usually doesn't result in the expected behavior. You can accomplish what you want using a tf.Variable. See [here](https://github.com/bgroenks96/normalizing-flows/blob/master/normalizing_flows/models/vae.py) for an example. This example is not with BNNs, so you can't literally copy-paste, but the method is transferrable. You need to define a tf.Variable, use it in your divergence function, and then update it during training.

**Comment by nbro** [comment]
@bgroenks96 Yes, I had already solved this problem by relying on variables. Thank you anyway!

**Comment by SouBanerjee** [comment]
@nbro can you post the solution here?
Please!
